### PR TITLE
fix(booking,finance): correctness, schema-freeze & DST de-scope

### LINKED ISSUE #440: Add DB-level exclusion constraint for consultant slot overlap prevention
## Summary

There is no database-level constraint preventing double-booked appointment slots for the same consultant. Currently, overlap prevention relies entirely on application-level checks (distributed Redis locks + Prisma queries), which are vulnerable to race conditions under concurrent writes.

## Problem

When two booking requests arrive simultaneously for the same consultant time slot, it's possible for both to pass the application-level overlap check before either commits, resulting in double-booked slots. The distributed lock mitigates this but doesn't fully eliminate the risk (e.g., lock expiry, network partitions).

## Proposed Solution

Add a PostgreSQL exclusion constraint on the `SlotOfAppointment` table using `btree_gist`:

### 1. Enable the `btree_gist` extension
```sql
CREATE EXTENSION IF NOT EXISTS btree_gist;
```

### 2. Add denormalized `consultantProfileId` column to `SlotOfAppointment`
```sql
ALTER TABLE "SlotOfAppointment"
ADD COLUMN "consultantProfileId" TEXT;
```

Populate from existing data via the appointment → event → plan chain, then add NOT NULL constraint.

### 3. Create exclusion constraint
```sql
ALTER TABLE "SlotOfAppointment"
ADD CONSTRAINT no_consultant_slot_overlap
EXCLUDE USING gist (
  "consultantProfileId" WITH =,
  tstzrange("startsAt", "endsAt") WITH &&
);
```

### 4. Update Prisma schema
Add the `consultantProfileId` field to the `SlotOfAppointment` model and add a relation to `ConsultantProfile`.

### 5. Update application code
- Set `consultantProfileId` when creating `SlotOfAppointment` records
- Handle the exclusion constraint violation error gracefully (Prisma error code P2002 or raw constraint violation)

## Why This Was Deferred

- Requires a Supabase migration + Prisma schema change + data backfill
- The `btree_gist` extension needs to be enabled in Supabase
- Denormalization adds write-time overhead and needs to stay in sync
- Too high-risk to bundle with the booking algorithm bug fix PR

## Related

- Part of booking algorithm bug review (Bug #16 in `bugs/booking-algorithm/16-missing-db-level-overlap-constraint.md`)
- Fixed in PR: application-level overlap checks improved (canonical overlap predicate, consistent lock keys, centralized occupancy policy)

## Labels
enhancement, database, booking-system

### LINKED ISSUE #502: Evaluate JS Temporal API migration — deferred (revisit late 2026)
## Context

JavaScript's [Temporal API](https://tc39.es/proposal-temporal/) reached **TC39 Stage 4** in March 2026 after 10 years of development. As a multi-country consultation platform handling dates across timezones globally, we evaluated whether migrating from our current `date-fns` + `date-fns-tz` stack to Temporal is worth the effort.

**Decision: Not now.** The ecosystem isn't ready for our stack, and our existing architecture already solves most problems Temporal addresses. This issue documents the full research for future reference.

---

## What Temporal API Solves (vs Legacy `Date`)

| Problem | `Date` Object | `Temporal` |
|---------|--------------|------------|
| **Timezones** | No native support, requires `date-fns-tz` | Built-in IANA timezone support |
| **DST Ambiguity** | Silent bugs when clocks fall back/forward | Explicit `disambiguation: "earlier" \| "later"` option |
| **Date Arithmetic** | Mutation-based, error-prone | Immutable, chainable operations |
| **Parsing** | Permissive, accepts invalid dates silently | Strict — throws on invalid input |
| **Calendar Systems** | Gregorian only | Multiple calendar support |
| **Immutability** | Mutable objects, side effects | Immutable by design |
| **Type Safety** | Loose type coercion | Prevents mixing incompatible types (PlainDate vs ZonedDateTime) |

### Key Temporal Types (relevant to our scheduling)

- **`Temporal.Instant`** — Exact UTC moment (for: event logs, token expiry, audit trails, DB storage)
- **`Temporal.ZonedDateTime`** — Timezone-aware datetime (for: meetings, user-facing times, "9am in user's timezone")
- **`Temporal.PlainDate`** — Date only (for: birthdays, deadlines, cutoff dates)
- **`Temporal.PlainTime`** — Time only (for: recurring availability slots — analogous to our SmallInt UTC minutes)
- **`Temporal.PlainDateTime`** — Local date+time, no timezone (for: UI forms before timezone is known)
- **`Temporal.Duration`** — Time spans (for: session lengths, hold periods)

---

## Current Temporal API Status (March 2026)

| Platform | Support | Notes |
|----------|---------|-------|
| **Chrome** | 144+ (Jan 2026) | Fully shipped |
| **Firefox** | 139+ (May 2025) | Fully shipped |
| **Safari** | Tech Preview ONLY | No confirmed release date (expected late 2026) |
| **Edge** | 144+ (Jan 2026) | Shipped |
| **Node.js 22** | `--harmony-temporal` flag only | NOT enabled by default |
| **TypeScript** | 6.0+ required | We're on 5.9.3 |
| **Prisma** | NO native support | [Issue #16119](https://github.com/prisma/prisma/issues/16119) open, not prioritized |
| **Polyfill** | `@js-temporal/polyfill` ~150KB | Production-usable but heavy |

---

### Scale of Date Handling
- **471+ files** with date handling logic
- **80+ DateTime fields** across 60+ Prisma models
- **4 timezone storage fields** (`User.timezone`, `Subscription.schedulingTimezone`, `Class.schedulingTimezone`, `NotificationPreference.quietHoursTimezone`)
- **SmallInt UTC-minutes pattern** for weekly availability (Int 0-1439 + DayOfWeek enum)
- **25+ cron jobs** with date-based scheduling
- **date-fns 4.1.0 + date-fns-tz 3.2.0** — sole date library

### Key Architecture Decision Already Made
Our `SlotOfAvailabilityWeekly` model stores availability as **UTC minutes since midnight** (`Int @db.SmallInt`) + `DayOfWeek` enum, rather than DateTime. This was a deliberate fix (Issue #6 / PR #446) that **already solves Temporal's biggest selling point** — DST-proof scheduling. The SmallInt representation is timezone-agnostic by design.

### 21 Risk Areas Identified

**HIGH RISK (7 areas, ~280 lines):**
1. Overnight slot logic — 3 representations, `isOvernightUTC` flag complexity (`utils/schedule/formatting.ts:174-177`)
2. `getTimezoneOffsetMinutes()` — fragile `toLocaleString()` heuristic (`utils/slotAllocation/slotTimeUtils.ts:307-326`)
3. Slot overlap detection — 3 separate OR scenarios in Prisma (`slotTimeUtils.ts:149-247`)
4. Custom slot midnight crossing — asymmetric `isEndTime` flag logic (`formatting.ts:235-269`)
5. Subscription end-date month overflow — Jan 31 + 1 month edge cases (`utils/dateUtils.ts:17-42`)
6. Weekly slot processing — complex `toZonedTime`/`fromZonedTime` roundtrip (`utils/timeSlotsProcessing.ts:81-176`)
7. 24-hour reschedule window — no DST adjustment (`app/api/appointments/[appointmentId]/reschedule/route.ts`)

**MEDIUM RISK (8 areas, ~400 lines):**
Checkout validation, SlotAllocationService, SlotValidationService, availability API timezone conversion, calendar expansion, cron job time thresholds, Stream recording sync, waitlist expiration

**LOW RISK (6 areas, ~150 lines):**
Display formatting, experience years, auth token expiry, feedback timestamps, payment history display, rate limiting

---

### What it would involve
- Rewrite 471+ files from `date-fns` to Temporal
- Build custom Prisma serialization layer for 80+ DateTime fields (Temporal → BigInt/String → Temporal on every read/write)
- Upgrade TypeScript 5.9.3 → 6.0+
- Use `--harmony-temporal` Node.js flag in production
- Add 150KB polyfill for Safari users
- Rewrite SmallInt minutes ↔ `Temporal.PlainTime` conversion layer
- Rewrite all 25+ cron jobs
- **Estimated time: 3-4 weeks minimum**

### Why NOT now
1. **Prisma blocker** — No native Temporal support means writing a serialization layer for EVERY database query. With 80+ DateTime fields across 60+ models, this is essentially building a mini-ORM adapter. It would be a new bug surface in itself.
2. **SmallInt minutes already works** — Our most critical scheduling logic (weekly availability) deliberately bypasses `Date` entirely. Temporal's `PlainTime` would provide the same DST-immunity we already have.
3. **No active bugs** — 6 rounds of E2E testing (agents 001-006) caught and fixed major date issues. We're solving a problem that doesn't exist yet.
4. **TypeScript 6.0 upgrade risk** — Major TS version bumps can break existing type inference, generic constraints, and third-party type definitions.
5. **Node.js flag risk** — Running `--harmony-temporal` in production for a pre-launch SaaS is risky. Non-default flags can have unexpected behavior.
6. **471+ files = massive blast radius** — Pre-launch is the worst time for a full rewrite of working code.

---

### What it would involve
- Install `@js-temporal/polyfill` for browser, use `--harmony-temporal` for server
- Write new features (post-launch) using Temporal types
- Harden existing date-fns code in parallel
- Build adapter functions at boundaries (Temporal ↔ Date ↔ Prisma)
- **Estimated time: 2 weeks initial + ongoing overhead**

### Why NOT now
1. **Two date systems in one codebase** — Every dev must know both `date-fns` patterns AND Temporal. With a 2-3 person team, this is cognitive overhead that doesn't pay off.
2. **Boundary bugs** — Converting between `Date` and `Temporal.Instant` at module boundaries introduces a new class of bugs.
3. **Still needs Prisma layer** — Any data touching Temporal needs the serialization workaround.
4. **"Where does date-fns end and Temporal begin?"** — Without clear module boundaries, this becomes a constant architectural question.
5. **Inconsistency** — Half the codebase uses one system, half uses another. PRs become harder to review.

---

## Recommended: Approach C — Harden date-fns (see #503)

A separate issue (#TBD) tracks the concrete technical implementation plan for hardening the 7 high-risk areas using the existing `date-fns` + `date-fns-tz` stack.

---

## When to Revisit Temporal (Trigger Conditions)

Reopen this evaluation when **ALL** of the following are true:

- [ ] **Prisma ships native Temporal support** — Watch [prisma/prisma#16119](https://github.com/prisma/prisma/issues/16119)
- [ ] **Node.js enables Temporal by default** — Expected in Node 24 LTS or later
- [ ] **Safari ships Temporal** — Expected late 2026, eliminates polyfill need
- [ ] **TypeScript 6.0+ is adopted** — We've upgraded and stabilized on TS 6
- [ ] **Production date bugs emerge** — Real-world traffic reveals issues that date-fns can't handle

**Estimated revisit window: Q4 2026 – Q1 2027**

---

## References

- [TC39 Temporal Proposal](https://tc39.es/proposal-temporal/)
- [MDN: Temporal API](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Temporal)
- [Prisma Issue #16119: TC39 Temporal Support](https://github.com/prisma/prisma/issues/16119)
- [Socket.dev: TC39 Advances Temporal to Stage 4](https://socket.dev/blog/tc39-advances-temporal-to-stage-4)
- [Bloomberg JS Blog: Temporal's 9-Year Journey](https://bloomberg.github.io/js-blog/post/temporal/)
- [Bryntum: JavaScript Temporal in 2026](https://bryntum.com/blog/javascript-temporal-is-it-finally-here/)
- Our E2E testing: Agents 001-006 (`prompts/booking-algorithm-tests/`)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### LINKED ISSUE #503: Harden date/timezone handling — 7 high-risk areas (Option C)
## Context

Following the Temporal API evaluation (#502), we decided to **keep our current `date-fns` + `date-fns-tz` stack** and harden the 7 high-risk areas identified during the audit. Our SmallInt UTC-minutes architecture is fundamentally sound after 6 rounds of E2E testing (agents 001-006), but these specific areas have fragile patterns that could cause issues at production scale with global users across DST boundaries.

**Estimated effort: 3-5 days**
**Priority: Before launch**

---

## Codebase Date Architecture (Current State)

| Layer | Approach | Status |
|-------|----------|--------|
| **DB: Weekly Slots** | `Int @db.SmallInt` (0-1439 UTC minutes) + `DayOfWeek` enum | Sound — DST-immune by design |
| **DB: Appointments/Events** | `DateTime @db.Timestamptz()` | Sound — PostgreSQL handles UTC storage |
| **DB: Timezone Storage** | `String` fields (`User.timezone`, `Subscription.schedulingTimezone`, etc.) | Sound |
| **App: Conversions** | `date-fns-tz` (`toZonedTime`/`fromZonedTime`) | Sound but some fragile heuristics |
| **App: Arithmetic** | `date-fns` (`addDays`, `startOfDay`, etc.) | Sound |
| **App: Overnight Detection** | Triple-representation (`isOvernightUTC` flag + local + UTC) | Fragile — needs consolidation |

**Key files:**
- `utils/dateTimeUtils.ts` (334 lines) — Core conversion utilities
- `utils/slotAllocation/slotTimeUtils.ts` (384 lines) — Slot math, overlap detection, offset calc
- `utils/schedule/formatting.ts` (366 lines) — UI ↔ API slot formatting
- `utils/timeSlotsProcessing.ts` (729 lines) — Weekly/custom slot expansion engine
- `utils/dateUtils.ts` (83 lines) — Subscription date math

---

### 1. `getTimezoneOffsetMinutes()` — Fragile Heuristic
**File:** `utils/slotAllocation/slotTimeUtils.ts:307-326`
**Risk:** Uses `toLocaleString()` to compute timezone offset — this is a **known unreliable pattern** that can produce wrong results during DST transitions and on some JS engines.

**Current code (fragile):**
```typescript
export function getTimezoneOffsetMinutes(timezone: string): number {
  const date = new Date();
  const utcStr = date.toLocaleString("en-US", { timeZone: "UTC" });
  const tzStr = date.toLocaleString("en-US", { timeZone: timezone });
  return Math.round((new Date(tzStr).getTime() - new Date(utcStr).getTime()) / 60000);
}
```

**Fix:** Replace with `date-fns-tz`'s proper IANA database lookup. `date-fns-tz` already uses the full IANA timezone database internally — we should leverage it instead of the `toLocaleString` heuristic.

**Impact:** Used in `isMinuteWithinWeeklySlot()` which validates checkout slots. Wrong offset = user books a slot 30-60 minutes off from their intended time.

---

### 2. Overnight Slot Triple-Representation
**Files:** `utils/schedule/formatting.ts:174-177`, `utils/slotAllocation/slotTimeUtils.ts:97-105, 259-299`
**Risk:** Three different ways to represent "overnight":
1. **Local time:** `endTime < startTime` (e.g., "23:00" → "02:00")
2. **UTC minutes:** `endMinutes < startMinutes` (e.g., 1320 → 120)
3. **`isOvernightUTC` flag:** Set when slot is overnight in UTC but same-day after local conversion

The `isOvernightUTC` flag was added during agent-005 E2E testing as a patch. The triple-representation makes it easy to double-count or miss overnight detection.

**Fix:** Create a single canonical `resolveOvernightStatus(slot, timezone)` function that returns a normalized `{ isOvernight: boolean, utcStartDay: DayOfWeek, utcEndDay: DayOfWeek }` object. All three files should call this single function instead of each computing overnight status independently.

**Impact:** Wrong overnight detection = slots appear on wrong day, or overlap detection misses a conflict → double-booking.

---

### 3. Slot Overlap Detection — 3 Separate OR Scenarios
**File:** `utils/slotAllocation/slotTimeUtils.ts:149-247`
**Risk:** `buildWeeklyOverlapWhere()` generates Prisma WHERE with 3 separate OR clauses for same-day overlap, carry-over from previous day, and starting-on-day overlap. Each clause has its own minute-range logic. Hard to reason about correctness.

**Fix:** Add comprehensive unit tests covering all 8 overlap edge cases:
1. Same-day full overlap
2. Same-day partial overlap (start inside existing)
3. Same-day partial overlap (end inside existing)
4. Same-day containment (new contains existing)
5. Overnight new overlaps same-day existing
6. Same-day new overlaps overnight existing
7. Overnight new overlaps overnight existing
8. Adjacent slots (no overlap — boundary case)

Also add a pure-function `slotsOverlapPure(slotA, slotB)` that can be unit-tested independently of Prisma, and validate it matches the Prisma query semantics.

**Impact:** Missed overlap = double-booking. False positive = user can't book valid slots.

---

### 4. Custom Slot Midnight Crossing
**File:** `utils/schedule/formatting.ts:235-269`
**Risk:** `convertTimezoneToUtcWithOvernight()` is called separately for start and end times, with an asymmetric `isEndTime` flag and `startTimeStr` parameter. The logic for detecting midnight crossing is split across two calls rather than computed holistically.

**Fix:** Refactor to compute start and end UTC times together in a single function call that can reason about the pair holistically. The function should accept `(startTime, endTime, date, timezone)` and return `{ utcStart: Date, utcEnd: Date }` with midnight crossing already resolved.

**Impact:** Off-by-one day on custom availability slots near midnight → user sees wrong date.

---

### 5. Subscription End-Date Month Overflow
**File:** `utils/dateUtils.ts:17-42`
**Risk:** `addMonthsSafely()` handles Jan 31 + 1 month → Feb 28 (not Mar 3). The logic is correct but has no test coverage. Edge cases: Feb 29 in leap years, Dec 31 + various months, month additions that span year boundaries.

**Fix:** Add unit tests for all edge cases:
- Jan 31 + 1 month = Feb 28 (non-leap) / Feb 29 (leap)
- Mar 31 + 1 month = Apr 30
- Dec 31 + 1 month = Jan 31 (next year)
- Jan 29 + 1 month in non-leap year = Feb 28
- Any date + 12 months = same date next year
- Start from Feb 29 (leap year) + 12 months = Feb 28 (non-leap)

**Impact:** Wrong subscription end date = billing errors. Revenue-critical.

---

### 6. Weekly Slot Processing — DST Transition Gap
**File:** `utils/timeSlotsProcessing.ts:81-176`
**Risk:** `processWeeklySlots()` iterates day-by-day using `addDays()` + `startOfDay()` in the target timezone, then extracts local time patterns and converts back to UTC. During **DST spring-forward**, `startOfDay()` + adding hours can land on a nonexistent local time (e.g., 2:30 AM doesn't exist when clocks jump 2:00 → 3:00). During **fall-back**, the same local time exists twice.

**Fix:** Add explicit DST guards:
1. After computing `fromZonedTime()`, verify the roundtrip: `toZonedTime(result, timezone)` should match the intended local time within 1-minute tolerance.
2. For spring-forward gaps: shift to the next valid time (e.g., 2:30 AM → 3:00 AM).
3. For fall-back ambiguity: pick the first occurrence (earlier UTC time).
4. Add a `DST_TRANSITION_TOLERANCE_MS` constant (60000ms) for comparison.

**Impact:** During DST transitions (2x per year per timezone), availability slots could be displayed at wrong times or missing entirely for that day.

---

### 7. 24-Hour Reschedule Window — No DST Handling
**File:** `app/api/appointments/[appointmentId]/reschedule/route.ts`
**Risk:** Uses `Date.getTime()` with `TWENTY_FOUR_HOURS_IN_MS` constant. During DST transitions, 24 clock hours ≠ 24 actual hours. A 23-hour DST day means users get 1 hour less than intended to reschedule.

**Fix:** Use `date-fns`'s `differenceInHours()` or `differenceInMilliseconds()` with timezone awareness, or accept that UTC-based 24-hour windows are intentional (document this decision explicitly). The fix depends on business intent:
- **If "24 hours before appointment start":** Current code is correct (UTC-based, DST-immune).
- **If "by the same time yesterday in user's timezone":** Needs `date-fns-tz` conversion.

**Action:** Confirm business intent and document the decision in a code comment. If UTC-based is intended, add a comment explaining why. If timezone-local is intended, refactor.

**Impact:** During DST transitions, users may be blocked from rescheduling 1 hour early or allowed 1 hour late. Affects ~4 days per year per timezone.

---

### Phase 1: Foundation (Day 1)
- [ ] Fix #1: Replace `getTimezoneOffsetMinutes()` with `date-fns-tz` proper offset lookup
- [ ] Fix #5: Add unit tests for `addMonthsSafely()` edge cases
- [ ] Fix #7: Confirm and document reschedule window business intent

### Phase 2: Overnight Consolidation (Day 2-3)
- [ ] Fix #2: Create canonical `resolveOvernightStatus()` function
- [ ] Fix #4: Refactor custom slot midnight crossing to use paired start/end computation
- [ ] Update `formatting.ts`, `slotTimeUtils.ts` to use the canonical function

### Phase 3: Robustness (Day 3-4)
- [ ] Fix #3: Add comprehensive overlap unit tests + pure-function validator
- [ ] Fix #6: Add DST transition guards to `processWeeklySlots()`

### Phase 4: Verification (Day 4-5)
- [ ] Run all 6 E2E booking test agents (001-006) against the changes
- [ ] Test with at least 3 timezone groups: IST (UTC+5:30), EST (UTC-5), AEDT (UTC+11)
- [ ] Test around DST transition dates (US: Mar 8 & Nov 1 2026, EU: Mar 29 & Oct 25 2026)
- [ ] Verify checkout flow end-to-end for all 4 service types

---

## Files to Modify

| File | Changes | Risk |
|------|---------|------|
| `utils/slotAllocation/slotTimeUtils.ts` | Replace offset heuristic, add overlap tests, consolidate overnight | High |
| `utils/schedule/formatting.ts` | Use canonical overnight function, refactor midnight crossing | High |
| `utils/timeSlotsProcessing.ts` | Add DST guards to processWeeklySlots | Medium |
| `utils/dateUtils.ts` | Add unit tests only (no code changes) | Low |
| `app/api/appointments/[appointmentId]/reschedule/route.ts` | Document or fix reschedule window | Low |
| New: `utils/schedule/overnight.ts` (or inline) | Canonical overnight resolution function | New |
| New: `tests/` or `__tests__/` | Unit tests for overlap, date math, DST | New |

---

## Verification Checklist

- [ ] All existing E2E tests pass (agents 001-006)
- [ ] New unit tests cover all 8 overlap scenarios
- [ ] `addMonthsSafely` tests cover leap year + month-end + year-boundary
- [ ] DST spring-forward: availability slot at 2:30 AM in US/Eastern on Mar 8 2026 resolves correctly
- [ ] DST fall-back: availability slot at 1:30 AM in US/Eastern on Nov 1 2026 resolves correctly
- [ ] Overnight slot IST 23:00-02:00 displays correctly for users in PST, EST, GMT, AEDT
- [ ] Checkout flow: book a consultation slot near midnight in 3 different timezones
- [ ] Subscription end-date: Jan 31 start + 1-month subscription shows Feb 28 end

---

## Related

- #502 — Temporal API evaluation (deferred)
- PR #446 — Original SmallInt minutes architecture fix
- Agent-005 — `isOvernightUTC` flag introduction
- Agent-006 — Lock contention and error classification fixes

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### PR DESCRIPTION
**Pull request by teetangh**
## Booking + finance productionization — #676/#677 tracker remainder, schema freeze, concurrency, DST de-scope

Hardens the (junior-written) booking + finance paths for launch. After a full enumeration of the **#676** (booking) and **#677** (payments) master-trackers, most P0/P1 findings were already fixed by prior PRs; this PR closes the genuine remaining tail plus the currency / DST / schema-freeze correctness work. Reviewed by adversarial passes (per-commit + whole-PR) and the CodeRabbit + Gemini bots — **all 13 review threads resolved**; `tsc` clean, full suites green.

### Already production-grade (confirmed, not rebuilt)
Postgres-native concurrency (ADR 13: Redis locks, status-CAS, Serializable+retry), the BigInt-paise double-entry ledger with a commit-time Σ(debit)==Σ(credit) trigger + INR-only guard (#783), comprehensive idempotency, wired rate-limiting, the live `slot_no_confirmed_overlap` exclusion constraint, and the no-broker posture (ADR 13/14; QStash deferred to #866).

### What changed
**Currency / money (settlement stays the `Currency` enum — ADR 15):** every ledger account is keyed INR (#783) — fixed at the org-payout settle + reversal legs, the wallet top-up, and the topup-refund webhook + its two readers; the class-multi reversal now carries the FX snapshot; `toCurrencyEnum` hardened (a blank gateway code dead-letters instead of coercing to INR); dead `|| "INR"` removed; non-INR manual invoices rejected until FX lands.

**Payments correctness (#677):** PM-12 (refund targets the *captured* payment), PM-11 (unique receipt), PM-4 (`CRITICAL_DISPUTE_UNLINKED` alert, paged once), PM-13 (no orphan FAILED refund), PM-15 + its `reconcile-payout-status` twin (both delegate to the canonical `handlePayoutWebhook` → TDS recorded, ledger counters fire, UTR persisted). #855 capture-after-cancel now auto-refunds (Phase-2, idempotent).

**Booking correctness (#676):** AE-1 (consultee-side conflict — the validator now scans `[consultant, consultee]`), RV-2 (validate/allocate agree on expired-payment slots via a shared helper that frees a hold only when *every* attempt is expired), RV-5 (class weekly-limit excludes the slots being replaced), AE-4 (`deletedAppointmentIds`), AE-5 (inlined slot duration). **#834** (waitlist tx) validated **already-fixed** — assignment runs inside the Serializable checkout tx and the m:n unique index is live.

**Concurrency:** CN-1 (event-checkout lock now fails **closed** on a Redis outage → typed 503 surfaced as a retryable 503, not a 500); CN-3 (deleted the never-wired event-slot semaphore and rewrote `12-concurrency-and-locking.md` to the real model — per-event mutex + a Serializable participant recount/P2034 for events; Read-Committed + the #440 exclusion constraint/`23P01`→409 for one-to-one slots).

**Schema freeze:** `@db.Timestamptz` on financial datetimes, `deletedAt` soft-delete, `AppointmentDocument.reviewedBy`→FK, custom-slot composite index, TDS challan/cert/ack fields, `ConsultantPayout.gatewayUtr` (closing a UTR-persistence gap), CHECK constraints (non-negative amounts, org-earnings split ≤ gross, financial-year regex), and `db:push` now auto-chains `db:constraints` (closing a silent-gap footgun). `UTR` kept as `gatewayUtr` (matches Razorpay's `utr` field) and documented.

**DST de-scope (ADR 17, #872):** pinned to IST; removed the half-built RFC-5545 materialization (`localTime.ts` + write-side), froze the IANA/local-wall-clock columns nullable so going DST-aware later needs no production migration; a missing timezone now defaults to IST (`330`).

**Error handling:** structured SQLSTATE predicates (`lib/db/pg-errors.ts`, unit-tested) replace message-substring matching; `23P01`/`23505` map to 409 at the source.

### Deferred to their own PRs/issues (with rationale)
- **AE-2** — class-collaborator availability (L-effort, highest regression surface).
- **PM-29** — refund-time FX-rate snapshot (needs a rate-source decision; #738-adjacent).
- **#860** — narrow the consultant-level auto-allocate lock (optimization; needs load-testing).
- **`withSerializableRetry` + P2034→409 on checkout** — the real last-seat contention-500 bug CN-3 surfaced (needs checkout-tx idempotency analysis + a load test).
- Larger schema decisions **A9 / A11 / B4** (explicit attendee join model, row-version column, `CancellationAuditLog`).
- A stale-semaphore **doc sweep** (5 non-booking docs) and a DRY-extract of the two payout reconcilers.
- **Capacity load-test** — run chaos scenario 6 at 2× peak (#874).

### Verification
`prisma validate` / `generate` pass; `tsc --noEmit` is 0 errors over the full branch; ~1100+ tests pass across booking / payments / enterprise / schedule / db with no weakened assertions. Adversarial reviews (per-commit + a whole-PR pass) came back clean, and all 13 CodeRabbit/Gemini threads are resolved.

Closes #503
Closes #502
Closes #440
Part of #837
Refs #676 #677 #781 #783 #860 #872 #874

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Checkout now shows targeted “pick another time” messages for booking conflicts and returns a retry-friendly “booking system busy” response if checkout locking is unavailable.
  * Dashboard planner/events refresh on browser window refocus.
* **Bug Fixes**
  * Expired pending-payment holds no longer block booking; partial reschedules now reflect freed tentative slots; weekly limits are enforced consistently.
  * Weekly availability no longer stores DST/local wall-clock fields.
  * Disputes/refunds/payouts avoid orphan records and persist the correct UTR; capture-after-terminal now triggers auto-refunds.
  * Wallet/ledger posting and enterprise payouts are INR-only; invoice creation rejects non-INR display currencies.
* **Refactor**
  * Centralized payment error handling with stronger conflict classification.
* **Documentation**
  * Updated currency, slot freshness, and IST launch decision docs.

### CODE REVIEW

File: utils/slotAllocation/SlotAllocationService.ts [hunk 1]
```diff
@@ -1406,45 +1407,58 @@ export class SlotAllocationService {
...
+            data: {
+              appointmentType: this.getAppointmentType(eventType),
+              [this.getEventRelationField(eventType)]: {
+                connect: { id: eventId },
+              },
+              ...(organizationId ? { organizationId } : {}),
+              // B1 — freeze the refund terms at booking (see cancellation-policy.ts).
+              cancellationPolicySnapshot: JSON.parse(
+                JSON.stringify(resolveCancellationPolicySnapshot()),
+              ),
```
[the remark this answers is not in the document]
Comment (teetangh) [reply utils/slotAllocation/SlotAllocationService.ts:1469]: Kept JSON.parse(JSON.stringify(...)): structuredClone throws in the jsdom test realm for this input (turned 32 suite tests red), and the snapshot is plain JSON (no Dates/Maps). Revisit with a Node test env. Addressed in 60152df2.

File: utils/slotAllocation/SlotAllocationService.ts [hunk 2]
```diff
@@ -1406,45 +1407,58 @@ export class SlotAllocationService {
...
-                ? [{ id: consultantUserId }, { id: consulteeUserId }]
-                : [{ id: consultantUserId }],
-            },
-          };
-        });
+    // Create appointment for each call. A concurrent booking that overlaps an
+    // existing confirmed slot trips the #440 exclusion constraint (or the unique
+    // guard); convert it to a typed 409 here at the source so classifyError can
+    // stay typed-only rather than sniffing Postgres error strings (#837).
+    let appointments: any[];
```
[the remark this answers is not in the document]
Comment (teetangh) [reply utils/slotAllocation/SlotAllocationService.ts:1440]: Kept any[]: the appointment.create({ include: { slotsOfAppointment } }) result does not narrow to AppointmentWithSlots[] through Promise.all + .map() (TS2322), and any[] matches the method Promise<any[]> return. Addressed in 60152df2.

File: app/api/slots/availability/weekly/route.ts [hunk 3]
```diff
@@ -178,13 +177,8 @@ export async function POST(req: NextRequest) {
...
       : 0;
-    // #503 — persist the DST-proof source of truth alongside the frozen
-    // offset; the slot math migrates read-side in the follow-up.
-    const timezone = consultantProfile.user?.timezone ?? null;
-    const localStartMinutes = toLocalMinutes(startTimeUtc, utcOffsetMinutes);
-    const localEndMinutes = toLocalMinutes(endTimeUtc, utcOffsetMinutes);
-    const localStartDay = toLocalDay(startDay, startTimeUtc, utcOffsetMinutes);
-    const localEndDay = toLocalDay(endDay, endTimeUtc, utcOffsetMinutes);
+    // TODO(#872): restore the local wall-clock + IANA-zone source of truth when
+    // non-IST consultants onboard; DST is parked post-MVP (IST-only at launch).
```
[the remark this answers is not in the document]
Comment (teetangh) [reply app/api/slots/availability/weekly/route.ts:181]: Applied the safe half: a missing timezone now defaults to IST 330 (not UTC 0) in both routes (60152df2). The hard reject of a non-IST timezone is a write-time product behavior change, deferred to #872 (DST/IANA-TZID implementation).

File: scripts/payouts/reconcile-payout-status.ts [hunk 4]
```diff
@@ -20,6 +20,22 @@
...
+  Record<
+    PayoutStatus,
+    "PENDING" | "PROCESSING" | "COMPLETED" | "FAILED" | "CANCELLED"
+  >
+> = {
+  [PayoutStatus.COMPLETED]: "COMPLETED",
+  [PayoutStatus.PROCESSING]: "PROCESSING",
+  [PayoutStatus.FAILED]: "FAILED",
+  [PayoutStatus.CANCELLED]: "CANCELLED",
+};
```
[the remark this answers is not in the document]
Comment (teetangh) [reply scripts/payouts/reconcile-payout-status.ts:38]: Acknowledged. The duplication across the two short reconcilers is intentional for now; extracting WEBHOOK_STATUS_MAP / getRazorpayPayoutStatus / mapGatewayStatus into a shared module is tracked as a follow-up refactor.
